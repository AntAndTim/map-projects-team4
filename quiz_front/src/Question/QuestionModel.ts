import {AnswerModel} from "../Answer/AnswerModel";

export class QuestionModel {
    constructor(id: number, text: string, answers: AnswerModel[]) {
        this.id = id;
        this.text = text;
        this.answers = answers;
    }

    id: number
    text: string
    answers: AnswerModel[]
}