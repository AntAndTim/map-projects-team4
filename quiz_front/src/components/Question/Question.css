.Question {
    width: 500px;
    display: flex;
    align-items: center;
    flex-direction: column;
    margin: auto;
}

.Question-Answers {
    display: flex;
    flex-direction: row;
    flex-wrap: wrap;

    justify-content: center;
}