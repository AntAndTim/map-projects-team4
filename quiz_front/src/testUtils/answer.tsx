export const getAnswer = (id: number, correct?: boolean, selected?: boolean) => ({
    id,
    text: `AnswerText${id}`,
    correct: Boolean(correct),
    selected: Boolean(selected),
})