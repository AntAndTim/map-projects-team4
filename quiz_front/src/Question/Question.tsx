import {Card} from "antd";
import {QuestionModel} from "./QuestionModel";
import {Answer} from "../Answer/Answer";
import React from "react";

export const Question = ({question}: { question: QuestionModel }) =>
    <Card title={question.text} style={{width: 300}}>
        <>
            {question.answers.map(answer => <Answer answer={answer}/>)}
        </>
    </Card>