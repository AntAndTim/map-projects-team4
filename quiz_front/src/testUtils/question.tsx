import {getAnswer} from "./answer";

export const getQuestion = (id: number) => ({
    id,
    text: `QuestionText${id}`,
    answers: [getAnswer(1), getAnswer(2), getAnswer(3, true)],
})