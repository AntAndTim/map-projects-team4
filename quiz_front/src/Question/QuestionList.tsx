import axios, {AxiosResponse} from "axios";
import {baseUrl} from "../config/Config";
import React, {useEffect, useState} from "react";
import {Question} from "./Question";
import {QuestionModel} from "./QuestionModel";
import {Col} from "antd";

export const QuestionList = () => {
    const [questions, setQuestions] = useState<QuestionModel[]>([])
    useEffect(() => {
        axios.get(baseUrl("question"))
            .then(function (response: AxiosResponse<QuestionModel[]>) {
                setQuestions(response.data);
            });
    }, [])

    return <>
        {questions.map((question: QuestionModel) =>
            <Col span={24 / questions.length}>
                <Question question={question}/>
            </Col>
        )}
    </>
}