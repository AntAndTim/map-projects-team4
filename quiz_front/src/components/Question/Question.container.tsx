import {QuestionModel} from "../../models/QuestionModel";
import React, {useCallback, useState} from "react";

import {AnswerModel} from "../../models/AnswerModel";
import {calculateColor} from "./Question.utils/Question.utils";
import {Question} from "./Question";

interface QuestionContainerProps {
    question: QuestionModel
}

export type AnswerClickHandler = (id: number) => void;
export type AnswerQuestion = () => void;

/**
 * Component for displaying question in quiz mode
 *
 * @component
 * @example
 * const answers = [{id: 100, text: 'apple', correct: true}, {id: 200, text: 'banana', correct: false}]
 * const question = {id: 121, text: 'Which fruit is green?', answers}
 *
 * return (
 *   <QuestionContainer
 *      question={question}
 *   />
 * )
 */
export const QuestionContainer: React.FC<QuestionContainerProps> = ({question}) => {
    const [answers, setAnswers] = useState<AnswerModel[]>(question.answers || []);
    const [haveAnswered, setHaveAnswered] = useState(false)

    const answerClickHandler = useCallback<AnswerClickHandler>((id: number) => {
        if (haveAnswered) {
            return;
        }

        setAnswers(answers => answers.map(answer => {
            if (id === answer.id) {
                return {...answer, selected: !answer.selected};
            }

            return answer;
        }));
    }, [setAnswers, haveAnswered]);
    const answerQuestion = useCallback<AnswerQuestion>(() => {
        setHaveAnswered(true);

        setAnswers(answers => answers.map(answer => {
            return {
                ...answer,
                color: calculateColor(haveAnswered, answer),
            }
        }))
    }, [setHaveAnswered, setAnswers]);

    return (
        <Question
            question={question}
            answers={answers}
            onAnswerClick={answerClickHandler}
            haveAnswered={haveAnswered}
            answerQuestion={answerQuestion}
        />
    )
}