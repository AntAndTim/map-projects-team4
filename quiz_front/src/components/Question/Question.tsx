import {AnswerContainer} from "../Answer/Answer.container";
import {Button} from "antd";
import React from "react";
import {QuestionModel} from "../../models/QuestionModel";
import {AnswerModel} from "../../models/AnswerModel";
import {AnswerClickHandler, AnswerQuestion} from "./Question.container";

import './Question.css';

interface QuestionProps {
    question: QuestionModel;
    answers: AnswerModel[];
    onAnswerClick: AnswerClickHandler;
    haveAnswered: boolean;
    answerQuestion: AnswerQuestion;
}

export const Question: React.FC<QuestionProps> = props => {
    const {
        question,
        answers,
        onAnswerClick,
        haveAnswered,
        answerQuestion,
    } = props;

    return (
        <div className="Question">
            <h1 className="Question-Text">{question.text}</h1>
            <div className="Question-Answers">
                {
                    answers.map(answer => (
                        <AnswerContainer
                            answer={answer}
                            onAnswerClick={onAnswerClick}
                            haveAnswered={haveAnswered}
                        />
                    ))
                }
            </div>
            <Button
                type="primary"
                onClick={answerQuestion}
            >
                Answer
            </Button>
        </div>
    )
}