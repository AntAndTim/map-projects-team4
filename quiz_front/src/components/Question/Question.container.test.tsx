import React from 'react';
import {render} from 'enzyme';
import {QuestionContainer} from "./Question.container";
import {getQuestion} from "../../testUtils/question";

describe('Question', () => {
    it('Should render component', () => {
        const wrapper = render(
            <QuestionContainer
                question={getQuestion(200)}
            />
        );

        expect(wrapper).toMatchSnapshot();
    })
});