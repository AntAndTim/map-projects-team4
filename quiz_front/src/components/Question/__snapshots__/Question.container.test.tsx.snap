// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Question Should render component 1`] = `
<div
  class="Question"
>
  <h1
    class="Question-Text"
  >
    QuestionText200
  </h1>
  <div
    class="Question-Answers"
  >
    <div
      class="ant-card ant-card-bordered ant-card-hoverable Answer"
    >
      <div
        class="ant-card-body"
      >
        AnswerText1
      </div>
    </div>
    <div
      class="ant-card ant-card-bordered ant-card-hoverable Answer"
    >
      <div
        class="ant-card-body"
      >
        AnswerText2
      </div>
    </div>
    <div
      class="ant-card ant-card-bordered ant-card-hoverable Answer"
    >
      <div
        class="ant-card-body"
      >
        AnswerText3
      </div>
    </div>
  </div>
  <button
    class="ant-btn ant-btn-primary"
    type="button"
  >
    <span>
      Answer
    </span>
  </button>
</div>
`;
